File: src/network/packet/impl/LanguageSyncPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl;

use JsonException;
use pocketcloud\cloud\language\Language;
use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\packet\ClientboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\util\PacketData;

final class LanguageSyncPacket extends CloudPacket implements ClientboundPacket {

    public function __construct(
        private readonly string $language = "",
        private readonly array $messages = []
    ) {}

    public function handle(ServerClient $client): void {}

    /**
     * @throws JsonException
     */
    public function encodePayload(PacketData $packetData): void {
        $packetData->writeAll($this->language, base64_encode(gzencode(json_encode($this->messages, JSON_THROW_ON_ERROR | JSON_INVALID_UTF8_IGNORE | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE))));
    }

    public function decodePayload(PacketData $packetData): void {}

    public function getLanguage(): string {
        return $this->language;
    }

    public function getMessages(): array {
        return $this->messages;
    }

    public static function create(string $language, array $messages): self {
        return new self($language, $messages);
    }

    public static function fromLanguage(?Language $language = null): self {
        $language = $language ?? Language::current();
        return new self($language->getName(), $language->getMessages());
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/network/packet/impl/LanguageSyncPacket.java
package de.pocketcloud.cloud.network.packet.impl;

import de.pocketcloud.cloud.language.Language;
import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.network.packet.ClientboundPacket;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.network.packet.data.PacketData;
import de.pocketcloud.common.util.FileUtils;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;
import java.util.zip.GZIPOutputStream;

@NoArgsConstructor
@Getter
public final class LanguageSyncPacket extends CloudPacket implements ClientboundPacket {

    private String language;
    private Map<String, String> messages;

    public LanguageSyncPacket(String language, Map<String, String> messages) {
        this.language = language != null ? language : "";
        this.messages = messages != null ? messages : Map.of();
    }

    @Override
    public void handle(@NotNull ServerClient client) {}

    @Override
    public void encodePayload(PacketData packetData) {
        try {
            byte[] compressed = gzipCompress(FileUtils.GSON.toJson(messages).getBytes(StandardCharsets.UTF_8));
            String encoded = Base64.getEncoder().encodeToString(compressed);
            packetData.writeAll(language, encoded);
        } catch (Exception e) {
            throw new RuntimeException("Failed to encode LanguageSyncPacket", e);
        }
    }

    @Override
    public void decodePayload(PacketData packetData) {}

    private byte[] gzipCompress(byte[] data) throws Exception {
        var baos = new ByteArrayOutputStream();
        try (var gzip = new GZIPOutputStream(baos)) {
            gzip.write(data);
        }
        return baos.toByteArray();
    }

    public static LanguageSyncPacket create(String language, Map<String, String> messages) {
        return new LanguageSyncPacket(language, messages);
    }

    public static LanguageSyncPacket fromLanguage(Language language) {
        Language lang = language != null ? language : Language.current();
        return new LanguageSyncPacket(lang.name(), lang.messages());
    }

    public static LanguageSyncPacket fromLanguage() {
        return fromLanguage(null);
    }
}
